sentry_sdk/profiler.py

```diff
@@ -13,200 +13,333 @@
+    largely not a problem.
+    """
+
+    def __init__(self, capacity):
+        # type: (int) -> None
+
+        self.buffer = [None] * capacity
+        self.capacity = capacity
+        self.idx = 0
+
+    def write(self, sample):
+        # type: (Any) -> None
         """
-        Exports this object to a JSON format compatible with Sentry's profiling visualizer.
-        Returns dictionary which can be serialized to JSON.
+        Writing to the buffer is not thread safe. There is the possibility
+        that parallel writes will overwrite one another.
+
+        This should only be a problem if the signal handler itself is
+        interrupted by the next signal.
+        (i.e. SIGPROF is sent again before the handler finishes).
+
+        For this reason, and to keep it performant, we've chosen not to add
+        any synchronization mechanisms here like locks.
         """
-        return {
-            "samples": [
-                {
-                    "frames": sample.stack,
-                    "relative_timestamp_ns": sample.sample_time,
-                    "thread_id": thread_id(),
-                }
-                for sample in self.stack_samples
-            ],
-            "frames": [
-                {
-                    "name": frame.function_name,
-                    "file": frame.file_name,
-                    "line": frame.line_number,
+        idx = self.idx
+        self.buffer[idx] = sample
+        self.idx = (idx + 1) % self.capacity
+
+    def slice_profile(self, start_ns, stop_ns):
+        # type: (int, int) -> Dict[str, List[Any]]
+        samples = []  # type: List[Any]
+        frames = dict()  # type: Dict[FrameData, int]
+        frames_list = list()  # type: List[Any]
+
+        # TODO: This is doing an naive iteration over the
+        # buffer and extracting the appropriate samples.
+        #
+        # Is it safe to assume that the samples are always in
+        # chronological order and binary search the buffer?
+        for raw_sample in self.buffer:
+            if raw_sample is None:
+                continue
+
+            ts = raw_sample[0]
+            if start_ns > ts or ts > stop_ns:
+                continue
+
+            for tid, stack in raw_sample[1]:
+                sample = {
+                    "frames": [],
+                    "relative_timestamp_ns": ts - start_ns,
+                    "thread_id": tid,
                 }
-                for frame in self.frame_list()
-            ],
-        }
 
-    def frame_list(self):
-        # type: () -> typing.List[FrameData]
-        # Build frame array from the frame indices
-        frames = [None] * len(self._frame_indices)  # type: typing.List[typing.Any]
-        for frame, index in self._frame_indices.items():
-            frames[index] = frame
-        return frames
+                for frame in stack:
+                    if frame not in frames:
+                        frames[frame] = len(frames)
```

Comment:
I couldn't follow what you're doing with the `frames` dict here.
Is it memoizing something? And you update the `frames` dict with the current length of the dict itself, or it that a typo?

Reply:
@sl0thentr0py this is format size optimization. You index the stack frames so that you convert each sample's `frame[]` stack to a `int[]` and store each frame in the dict. The len(frames) is the sequential id of each frame which is then resolved when we reconstruct the stack.